File: NetDeck/CardManager.m
//
//  CardManager.m
//  Net Deck
//
//  Created by Gereon Steffens on 22.06.14.
//  Copyright (c) 2015 Gereon Steffens. All rights reserved.
//

#import "CardManager.h"
#import "SettingsKeys.h"
#import "CardSets.h"
#import "ImageCache.h"
#import "AppDelegate.h"
#import <DTCoreText.h>

@implementation CardManager

static NSMutableArray* allRunnerCards;          // all non-identity runner cards
static NSMutableArray* allCorpCards;            // all non-identity corp cards
static NSMutableArray* allRunnerIdentities;     // all runner ids
static NSMutableArray* allCorpIdentities;       // all corp ids

static NSArray* subtypes;       // array[role] of dictionary type->array
static NSArray* identitySubtypes; // array[role] of set of strings
static NSString* identityKey;

static NSMutableArray* sortedIdentities;

static NSMutableDictionary* allCards;   // code -> card

static int maxMU;
static int maxStrength;
static int maxRunnerCost;
static int maxCorpCost;
static int maxInf;
static int maxAgendaPoints;
static int maxTrash;

static NSDictionary* cardAliases;   // code -> alias

+(void) initialize
{
    allCards = [NSMutableDictionary dictionary];
    
    allRunnerCards = [NSMutableArray array];
    allCorpCards = [NSMutableArray array];
    
    allRunnerIdentities = [NSMutableArray array];
    allCorpIdentities = [NSMutableArray array];
    
    subtypes = @[ [NSMutableDictionary dictionary], [NSMutableDictionary dictionary] ];
    identitySubtypes = @[ [NSMutableSet set], [NSMutableSet set] ];
    sortedIdentities = [@[ [NSMutableArray array], [NSMutableArray array] ] mutableCopy];
    
    cardAliases = @{
        @"08034": @"Franklin",  // crick
        @"02085": @"HQI",       // hq interface
        @"02107": @"RDI",       // r&d interface
        @"06033": @"David",     // d4v1d
        @"05039": @"SW35",      // unreg. s&w '35
        @"03035": @"LARLA",     // levy ar lab access
        @"04029": @"PPVP",      // prepaid voicepad
        @"01092": @"SSCG",      // sansan city grid
        @"04034": @"SFSS",      // shipment from sansan
        @"03049": @"ProCo",     // professional contacts
        @"02079": @"OAI",       // oversight AI
        @"08009": @"Baby",      // symmetrical visage
        @"08003": @"Pancakes",  // adjusted chronotype
        @"09022": @"ASI",       // the all-seeing i
    };
}

+(Card*) cardByCode:(NSString*)code
{
    return [allCards objectForKey:code];
}

+(NSArray*) allCards
{
    NSMutableArray* cards = allCards.allValues.mutableCopy;

    return [cards sortedArrayUsingComparator:^(Card* c1, Card *c2) {
        NSUInteger l1 = c1.name.length;
        NSUInteger l2 = c2.name.length;
        if (l1 > l2)
        {
            return NSOrderedAscending;
        }
        else if (l1 < l2)
        {
            return NSOrderedDescending;
        }
        return NSOrderedSame;
    }];
}

+(NSArray*) allForRole:(NRRole)role
{
    return role == NRRoleRunner ? allRunnerCards : allCorpCards;
}

+(NSArray*) identitiesForRole:(NRRole)role
{
    return role == NRRoleRunner ? allRunnerIdentities : allCorpIdentities;
}

+(NSArray*) allRunnerCards
{
    return allRunnerCards;
}

+(NSArray*) allCorpCards
{
    return allCorpCards;
}

#pragma mark subtypes

+(NSMutableArray*) subtypesForRole:(NRRole)role andType:(NSString*)type includeIdentities:(BOOL)includeIds
{
    NSMutableArray* arr = [NSMutableArray arrayWithArray:subtypes[role][type]];
    
    includeIds = includeIds && ([type isEqualToString:kANY] || [type isEqualToString:identityKey]);
    if (includeIds)
    {
        if (!arr)
        {
            arr = [NSMutableArray array];
        }
        NSSet* set = identitySubtypes[role];
        for (NSString* s in set)
        {
            [arr addObject:s];
        }
    }
    
    if (arr)
    {
        [arr sortUsingSelector:@selector(localizedCaseInsensitiveCompare:)];
        return arr;
    }
    return nil;
}

+(NSMutableArray*) subtypesForRole:(NRRole)role andTypes:(NSSet*)types includeIdentities:(BOOL)includeIds
{
    NSMutableSet* subtypes = [NSMutableSet set];
    for (NSString* type in types)
    {
        NSMutableArray* arr = [NSMutableArray arrayWithArray:[CardManager subtypesForRole:role andType:type includeIdentities:includeIds]];
        if (arr.count > 0)
        {
            [subtypes addObjectsFromArray:arr];
        }
    }
    
    return [[[subtypes allObjects] sortedArrayUsingSelector:@selector(localizedCaseInsensitiveCompare:)] mutableCopy];
}

#pragma mark initalization

+(BOOL) cardsAvailable
{
    return allRunnerCards.count > 0 && allCorpCards.count > 0;
}

+(NSString*) filename
{
    NSArray* paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
    NSString* supportDirectory = [paths objectAtIndex:0];
    
    return [supportDirectory stringByAppendingPathComponent:CARDS_FILENAME];
}

+(NSString*) filenameEn
{
    NSArray* paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
    NSString* supportDirectory = [paths objectAtIndex:0];
    
    return [supportDirectory stringByAppendingPathComponent:CARDS_FILENAME_EN];
}

+(void) removeFiles
{
    NSFileManager* fileMgr = [NSFileManager defaultManager];
    [fileMgr removeItemAtPath:[CardManager filename] error:nil];
    [fileMgr removeItemAtPath:[CardManager filenameEn] error:nil];
    
    [CardManager initialize];
}

+(BOOL) setupFromFiles
{
    NSString* cardsFile = [CardManager filename];
    NSString* cardsEnFile = [CardManager filenameEn];
    BOOL ok = NO;
    
    NSFileManager* fileMgr = [NSFileManager defaultManager];
    if ([fileMgr fileExistsAtPath:cardsFile])
    {
        NSArray* data = [NSArray arrayWithContentsOfFile:cardsFile];
        if (data)
        {
            ok = [self setupFromJsonData:data];
        }
    }
    
    if (ok && [fileMgr fileExistsAtPath:cardsEnFile])
    {
        NSArray* data = [NSArray arrayWithContentsOfFile:cardsEnFile];
        if (data)
        {
            [self addAdditionalNames:data saveFile:NO];
        }
    }
    
    return ok;
}

+(BOOL) setupFromNrdbApi:(NSArray*)json
{
    [CardManager setNextDownloadDate];
    
    NSString* cardsFile = [CardManager filename];
    [json writeToFile:cardsFile atomically:YES];
    [AppDelegate excludeFromBackup:cardsFile];
    
    [CardManager initialize];
    return [self setupFromJsonData:json];
}

+(void) setNextDownloadDate
{
    NSDateFormatter *fmt = [NSDateFormatter new];
    [fmt setDateStyle:NSDateFormatterShortStyle]; // e.g. 08.10.2008 for locale=de
    [fmt setTimeStyle:NSDateFormatterNoStyle];
    NSDate* now = [NSDate date];
    
    NSUserDefaults* settings = [NSUserDefaults standardUserDefaults];
    [settings setObject:[fmt stringFromDate:now] forKey:LAST_DOWNLOAD];
    
    NSInteger interval = [settings integerForKey:UPDATE_INTERVAL];

    NSString* nextDownload;
    switch (interval)
    {
        case 30: {
            NSCalendar *cal = [NSCalendar currentCalendar];
            NSDate *next = [cal dateByAddingUnit:NSCalendarUnitMonth value:1 toDate:[NSDate date] options:0];
            nextDownload = [fmt stringFromDate:next];
            break;
        }
        case 0:
            nextDownload = l10n(@"never");
            break;
        default: {
            NSDate* next = [NSDate dateWithTimeIntervalSinceNow:interval*24*60*60];
            nextDownload = [fmt stringFromDate:next];
            break;
        }
    }
    
    [settings setObject:nextDownload forKey:NEXT_DOWNLOAD];
}

+(void) addAdditionalNames:(NSArray *)json saveFile:(BOOL)saveFile
{
    // add english names from json
    if (saveFile)
    {
        NSString* cardsFile = [CardManager filenameEn];
        [json writeToFile:cardsFile atomically:YES];
        
        [AppDelegate excludeFromBackup:cardsFile];
    }
    
    for (NSDictionary* obj in json)
    {
        NSString* code = obj[@"code"];
        NSString* name_en = obj[@"title"];
        NSString* subtype = obj[@"subtype"];
        
        Card* card = [Card cardByCode:code];
        if (card)
        {
            card.name_en = [name_en stringByReplacingHTMLEntities];
            [card setAlliance:subtype];
            [card setVirtual:subtype];
        }
    }
    
    // add automatic aliases like "Self Modifying Code" -> "SMC"
    for (Card* card in allCards.allValues)
    {
        NSString* regexPattern = @"^|[- ]\\w";
        NSRegularExpression* regex = [[NSRegularExpression alloc] initWithPattern:regexPattern options:0 error:nil];
        
        NSMutableString* alias = [NSMutableString string];
        NSArray* matches = [regex matchesInString:card.name options:0 range:NSMakeRange(0, card.name.length)];
        for (NSTextCheckingResult* match in matches)
        {
            NSRange range = match.range;
            range.length = 1;
            if (range.location > 0)
            {
                range.location++;
            }
            NSString* ch = [card.name substringWithRange:range];
            [alias appendString:ch];
        }
        
        if (card.name.length > 2 && alias.length > 1)
        {
            card.alias = alias;
        }
        
        // NSLog(@"%@ -> %@", card.name, alias);
    }
    
    // add hard-coded aliases
    for (NSString* code in cardAliases.allKeys)
    {
        Card* card = [Card cardByCode:code];
        card.alias = [cardAliases objectForKey:code];
    }
}

+(BOOL) setupFromJsonData:(NSArray*)json
{
    if (json)
    {
        for (NSDictionary* obj in json)
        {
            Card* card = [Card cardFromJson:obj];
            if (card)
            {
                NSAssert(card.isValid, @"invalid card from %@", obj);
            
                [CardManager addCard:card];
            }
        }
                
        NSArray* cards = [allCards allValues];
        [Faction initializeFactionNames:cards];
        [CardType initializeCardTypes:cards];
        
        // sort identities by faction and name
        for (NSMutableArray* arr in @[ allRunnerIdentities, allCorpIdentities ])
        {
            [arr sortUsingComparator:^(Card* c1, Card* c2) {
                if (c1.faction > c2.faction) {
                    return NSOrderedDescending;
                }
            
                if (c1.faction < c2.faction) {
                    return NSOrderedAscending;
                }
                return [c1.name compare:c2.name];
            }];
        }
        
        return YES;
    }
    else
    {
        return NO;
    }
}

+(void) addCard:(Card*)card
{
    // add to dictionaries/arrays
    [allCards setObject:card forKey:card.code];
    
    if (card.type == NRCardTypeIdentity)
    {
        NSMutableArray* arr = card.role == NRRoleRunner ? allRunnerIdentities : allCorpIdentities;
        [arr addObject:card];
    }
    else
    {
        NSMutableArray* arr = card.role == NRRoleRunner ? allRunnerCards : allCorpCards;
        [arr addObject:card];
    }
    
    // calculate max values for filter sliders
    maxMU = MAX(card.mu, maxMU);
    maxTrash = MAX(card.trash, maxTrash);
    maxStrength = MAX(card.strength, maxStrength);
    maxInf = MAX(card.influence, maxInf);
    maxAgendaPoints = MAX(card.agendaPoints, maxAgendaPoints);
    if (card.role == NRRoleRunner)
    {
        maxRunnerCost = MAX(card.cost, maxRunnerCost);
    }
    else
    {
        maxCorpCost = MAX(card.cost, maxCorpCost);
    }
    
    // fill subtypes per role
    if (card.subtype)
    {
        // NSLog(@"%@", card.subtype);
        if (card.type == NRCardTypeIdentity)
        {
            identityKey = card.typeStr;
            NSMutableSet* set = identitySubtypes[card.role];
            for (NSString* st in card.subtypes)
            {
                [set addObject:st];
            }
        }
        else
        {
            NSMutableDictionary* dict = subtypes[card.role];
            
            if (dict[card.typeStr] == nil)
            {
                dict[card.typeStr] = [NSMutableArray array];
            }
            if (dict[kANY] == nil)
            {
                dict[kANY] = [NSMutableArray array];
            }
            for (NSString* st in card.subtypes)
            {
                for (NSMutableArray* arr in @[ dict[card.typeStr], dict[kANY]])
                {
                    if (![arr containsObject:st])
                    {
                        [arr addObject:st];
                    }
                }
            }
        }
    }
}

#pragma mark max values

+(int) maxMU { return maxMU; }
+(int) maxStrength { return maxStrength; }
+(int) maxInfluence { return maxInf; }
+(int) maxRunnerCost { return maxRunnerCost; }
+(int) maxCorpCost { return maxCorpCost; }
+(int) maxAgendaPoints { return maxAgendaPoints; }
+(int) maxTrash { return maxTrash; }

@end

File: NetDeck/CardManager.swift
//
//  CardManager.swift
//  NetDeck
//
//  Created by Gereon Steffens on 21.11.15.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import Foundation
import SwiftyUserDefaults

final class CardManager {
    
    static let cardsFilename = "nrcards2.json"
    
    private static var allCardsByRole = [Role: [Card] ]()    // non-id cards
    private static var allIdentitiesByRole = [Role: [Card] ]()    // ids
    private(set) static var quantities = [String: Int]()   // map of code+packCode -> quantity
    
    private static var allSubtypes = [Role: [String: Set<String> ] ]()
    private static var identitySubtypes = [ Role: Set<String> ]()
    private static var identityKey = ""
    
    private static var allKnownCards = [ String: Card ](minimumCapacity: 1600)
    
    private(set) static var maxMU: Int = -1
    private(set) static var maxInfluence: Int = -1
    private(set) static var maxStrength: Int = -1
    private(set) static var maxAgendaPoints: Int = -1
    private(set) static var maxRunnerCost: Int = -1
    private(set) static var maxCorpCost: Int = -1
    private(set) static var maxTrash: Int = -1
    
    static func initialize() {
        allKnownCards.removeAll()
        
        allCardsByRole[.runner] = { var c = [Card](); c.reserveCapacity(800); return c }()
        allCardsByRole[.corp] = { var c = [Card](); c.reserveCapacity(800); return c }()
        
        allIdentitiesByRole[.runner] = { var c = [Card](); c.reserveCapacity(50); return c }()
        allIdentitiesByRole[.corp] = { var c = [Card](); c.reserveCapacity(50); return c }()
        
        identitySubtypes[.runner] = Set<String>()
        identitySubtypes[.corp] = Set<String>()
        
        allSubtypes[.runner] = [:]
        allSubtypes[.corp] = [:]
        
        quantities = [:]
        
        maxMU = -1
        maxInfluence = -1
        maxStrength = -1
        maxAgendaPoints = -1
        maxRunnerCost = -1
        maxCorpCost = -1
        maxTrash = -1
    }
    
    static func maxCost(for role: Role) -> Int {
        switch role {
        case .none: return max(maxCorpCost, maxRunnerCost)
        case .runner: return maxRunnerCost
        case .corp: return maxCorpCost
        }
    }

    static func cardBy(_ code: String, useReplacements: Bool = true) -> Card? {
        return allKnownCards[code]
    }
    
    static func allCards() -> [Card] {
        let cards = allKnownCards.values
        
        return cards.sorted {
            return $0.name.count > $1.name.count
        }
    }
    
    static func allFor(_ role: Role) -> [Card] {
        if role != .none {
            return allCardsByRole[role]!
        } else {
            return allCardsByRole[.runner]! + allCardsByRole[.corp]!
        }
    }
    
    static func identitiesFor(_ role: Role) -> [Card] {
        assert(role != .none)
        return allIdentitiesByRole[role]!
    }
    
    static func identitiesForSelection(_ role: Role, packUsage: PackUsage, legality: DeckLegality) -> TableData<Card> {
        var factionNames = Faction.factionsFor(role: role, packUsage: packUsage)
        factionNames.removeFirst(2) // remove "any" and "neutral"
        
        let disabledPackCodes: Set<String>
        switch packUsage {
        case .all:
            var draft = Set<String>()
            if Defaults[.useDraft] {
                draft.insert(PackManager.draft)
            }
            disabledPackCodes = PackManager.rotatedPackCodes().union(draft)
        case .selected:
            disabledPackCodes = PackManager.disabledPackCodes().union(PackManager.rotatedPackCodes())
        }
        
        if !disabledPackCodes.contains(PackManager.draft) || Defaults[.useNapd] {
            factionNames.append(Faction.name(for: .neutral))
        }
        
        var identities = [[Card]]()
        factionNames.forEach { str in
            identities.append([])
        }

        var allIdentities = self.identitiesFor(role).filter { !disabledPackCodes.contains($0.packCode) }
        if packUsage == .selected && legality != .cacheRefresh && legality != .modded {
            for identity in Prebuilt.identities(for: role) {
                if !allIdentities.contains(identity) && !disabledPackCodes.contains(identity.packCode) {
                    allIdentities.append(identity)
                }
            }
        }

        if legality.isStandard {
            allIdentities = allIdentities.filter { !$0.banned(legality.mwl) }
        }

        allIdentities.sort { c1, c2 in
            c1.packNumber == c2.packNumber ? c1.number < c2.number : c1.packNumber < c2.packNumber
        }

        for identity in allIdentities {
            let faction = Faction.name(for: identity.faction)
            if let index = factionNames.firstIndex(where: { $0 == faction }) {
                identities[index].append(identity)
            }
        }

        if !disabledPackCodes.contains(PackManager.draft) {
            factionNames.removeLast()
            factionNames.append("Neutral/Draft".localized())
        }
        
        for i in (0 ..< factionNames.count).reversed() {
            if identities[i].count == 0 {
                identities.remove(at: i)
                factionNames.remove(at: i)
            }
        }

        assert(factionNames.count == identities.count)
        
        return TableData(sections: factionNames, values: identities)
    }
    
    static func subtypesFor(role: Role, andType type: String, includeIdentities: Bool) -> [String] {
        var subtypes = allSubtypes[role]?[type] ?? Set<String>()
        
        let includeIds = includeIdentities && (type == Constant.kANY || type == identityKey)
        if includeIds {
            if let set = identitySubtypes[role] {
                subtypes.formUnion(set)
            }
        }
    
        return subtypes.sorted { $0.lowercased() < $1.lowercased() }
    }
    
    static func subtypesFor(role: Role, andTypes types: Set<String>, includeIdentities: Bool) -> [String] {
        var subtypes = Set<String>()
        for type in types {
            let arr = subtypesFor(role: role, andType: type, includeIdentities: includeIdentities)
            subtypes.formUnion(arr)
        }
        if types.count == 0 {
            subtypes = (allSubtypes[role]?[Constant.kANY])!
        }
    
        return subtypes.sorted(by: { $0.lowercased() < $1.lowercased() })
    }
    
    static var cardsAvailable: Bool {
        return allKnownCards.count > 0 && PackManager.packsAvailable
    }
    
    private static func setSubtypes(_ cards: [Card]) {
        // fill subtypes per role
        for card in cards {
            if card.subtypes.count > 0 {
                // NSLog(@"%@", card.subtype)
                if card.type == .identity {
                    identityKey = card.typeStr
                    identitySubtypes[card.role]?.formUnion(card.subtypes)
                } else {
                    var dict = allSubtypes[card.role] ?? [String: Set<String>]()
                    
                    if dict[card.typeStr] == nil {
                        dict[card.typeStr] = Set<String>()
                    }
                    if dict[Constant.kANY] == nil {
                        dict[Constant.kANY] = Set<String>()
                    }
                    
                    dict[card.typeStr]?.formUnion(card.subtypes)
                    dict[Constant.kANY]?.formUnion(card.subtypes)
                    allSubtypes[card.role] = dict
                }
            }
        }
    }
    
    private static func addCardAliases(_ cards: [Card]) {
        // add automatic aliases like "Self Modifying Code" -> "SMC"
        let split = CharacterSet(charactersIn: " -.\"")

        cards.forEach { card in
            let words = card.name.components(separatedBy: split)
            if words.count > 1 {
                let alias = words.map { $0.prefix(1) }.joined(separator: "")
                card.addCardAlias(alias)
            }
        }

        // add hard-coded aliases
        for (code, alias) in Card.aliases {
            if let card = allKnownCards[code] {
                card.addCardAlias(alias)

                // find other cards with the same name
                let duplicates = allKnownCards.values.filter { $0.name == card.name && $0.code != card.code }
                for dup in duplicates {
                    allKnownCards[dup.code]?.addCardAlias(alias)
                }
            }
        }
    }
    
    private static func add(card: Card) {
        guard card.isValid else {
            print("invalid card: \(card.code) \(card.name)")
            return
        }
        
        allKnownCards[card.code] = card
        if card.type == .identity {
            allIdentitiesByRole[card.role]!.append(card)
        } else {
            allCardsByRole[card.role]!.append(card)
        }
        
        // calculate max values for filter sliders
        maxMU = max(card.mu, maxMU)
        maxTrash = max(card.trash, maxTrash)
        maxStrength = max(card.strength, maxStrength)
        maxInfluence = max(card.influence, maxInfluence)
        maxAgendaPoints = max(card.agendaPoints, maxAgendaPoints)
        
        if card.role == .runner {
            maxRunnerCost = max(card.cost, maxRunnerCost)
        } else {
            maxCorpCost = max(card.cost, maxCorpCost)
        }
    }
    
    static func setNextDownloadDate() {
        let fmt = DateFormatter()
        fmt.dateStyle = .short // e.g. 08.10.2008 for locale=de
        fmt.timeStyle = .none
        let now = Date()
        
        let interval = Defaults[.updateInterval]
        let nextDownload: String
        switch interval {
        case 30:
            let cal = Calendar.current
            let next = cal.date(byAdding: .month, value: 1, to: now)
            nextDownload = fmt.string(from: next!)
        case 0:
            nextDownload = "never".localized()
        default:
            let next = Date(timeIntervalSinceNow:TimeInterval(interval * 24*60*60))
            nextDownload = fmt.string(from: next)
        }

        Defaults[.lastDownload] = fmt.string(from: now)
        Defaults[.nextDownload] = nextDownload
        
        // print("next download \(nextDownload)")
    }
    
    // MARK: - persistence 
    
    static func setupFromFiles() -> Bool {
        let cardsFile = CardManager.filename()

        let fileMgr = FileManager.default

        if !fileMgr.fileExists(atPath: cardsFile) {
            // copy the file from our bundle
            if let bundlePath = Bundle.main.path(forResource: "cards_en", ofType: "json") {
                do {
                    try fileMgr.copyItem(atPath: bundlePath, toPath: cardsFile)
                } catch {
                    print(error)
                }
            }
        }

        if let data = FileManager.default.contents(atPath: cardsFile) {
            do {
                let decoder = JSONDecoder()
                let rawCards = try decoder.decode(ApiResponse<NetrunnerDbCard>.self, from: data)
 
                return setupFromJson(rawCards)
            } catch let error {
                print("\(error)")
                return false
            }
        }
        // print("app start: missing card file")
        return false
    }
    
    static func setupFromNetrunnerDb(_ cardsData: Data) -> Bool {
        var ok = false
        do {
            let decoder = JSONDecoder()
            let rawCards = try decoder.decode(ApiResponse<NetrunnerDbCard>.self, from: cardsData)
            ok = setupFromJson(rawCards)
            if !ok {
                return false
            }
            
            let filename = self.filename()
            try cardsData.write(to: URL(fileURLWithPath: filename), options: .atomic)
            Utils.excludeFromBackup(filename)
        } catch let error {
            print("\(error)")
        }
        
        return ok
    }
    
    private static func setupFromJson(_ rawCards: ApiResponse<NetrunnerDbCard>) -> Bool {
        if !rawCards.valid {
            return false
        }
        CardManager.initialize()
        
        // parse data
        let parsedCards = Card.cardsFromJson(rawCards)
        for card in parsedCards {
            CardManager.add(card: card)
        }

        let cards = Array(allKnownCards.values)
        if cards.count == 0 {
            return false
        }

        CardManager.setSubtypes(cards)
        CardManager.addCardAliases(cards)

        let factionsDict = cards.reduce(into: [Faction: String]()) { $0[$1.faction] = $1.factionStr }
        var ok = Faction.initializeFactionNames(factionsDict)
        if !ok {
            return false
        }

        let typesDict = cards.reduce(into: [CardType: String]()) { $0[$1.type] = $1.typeStr }
        ok = CardType.initializeCardType(typesDict)
        if !ok {
            return false
        }
        
        // sort identities by faction and name
        for var arr in [ allIdentitiesByRole[.runner]!, allIdentitiesByRole[.corp]! ] {
            arr.sort(by: { (c1, c2) -> Bool in
                if c1.faction.rawValue < c2.faction.rawValue { return true }
                if c1.faction.rawValue > c2.faction.rawValue { return false }
                return c1.name < c2.name
            })
        }

        #warning("GENERATE MAPPING")
//        let newCards = cards.filter { $0.packCode == "sg" || $0.packCode == "su21" }
//
//        var convert = [(Card, Card)]()
//        for card in cards {
//            if card.packCode == "sg" || card.packCode == "su21" { continue }
//
//            if let newCard = newCards.first(where: { $0.name == card.name }) {
//                convert.append((card, newCard))
//            }
//        }
//
//        convert.sort { c1, c2 in
//            if c1.0.name == c2.0.name {
//                return c1.0.code < c2.0.code
//            }
//            return c1.0.name < c2.0.name
//        }
//        print("let oldCoresToSU21 = [")
//        for (old, new) in convert {
//            print("    \"\(old.code)\": \"\(new.code)\" // \(old.name) - \(old.packName)")
//        }
//        print("]")
        
        return true
    }
    
    static func fileExists() -> Bool {
        return FileManager.default.fileExists(atPath: filename())
    }
    
    private static func filename() -> String {
        let paths = NSSearchPathForDirectoriesInDomains(.applicationSupportDirectory, .userDomainMask, true)
        let supportDirectory = paths[0]
    
        return supportDirectory.appendPathComponent(CardManager.cardsFilename)
    }
    
    static func removeFiles() {
        let fileMgr = FileManager.default
        _ = try? fileMgr.removeItem(atPath: filename())
    
        CardManager.initialize()
    }
}
